index.js

```diff
@@ -2095,11 +2096,38 @@
     };
   }
 
-  //  showType :: Type -> String
-  function showType(t) {
-    return unless(t.type === FUNCTION || t.type === RECORD || isEmpty(t.keys),
-                  stripOutermostParens,
-                  String(t));
+  //  typeVarNames :: Type -> Array String
+  function typeVarNames(t) {
+    return Z.concat(
+      t.type === VARIABLE ? [t.name] : [],
+      Z.chain(function(k) { return typeVarNames(t.types[k].type); }, t.keys)
+    );
+  }
+
+  //  typeVarChars :: Array String
+  var typeVarChars = (function() {
+    var chars = [];
+    var a = 'a'.charCodeAt(0);
+    var z = 'z'.charCodeAt(0);
+    for (var code = a; code <= z; code += 1) {
+      chars.push(String.fromCharCode(code));
+    }
+    return chars;
+  }());
+
+  //  showType :: TypeInfo -> Type -> String
+  function showType(typeInfo) {
+    return function(t) {
+      var names = Z.chain(typeVarNames, typeInfo.types);
+      var $free = typeVarChars.filter(function(name) {
+        return names.indexOf(name) < 0;
+      });
+      return unless(
+        t.type === FUNCTION || t.type === RECORD || isEmpty(t.keys),
+        stripOutermostParens,
+        String(t).replace(/\bUnknown\b/g, function() { return $free.shift(); })
```

Comment:
If you use `$free[idx++]` instead, you can also hoist the computation for `$free`.